artifact/lexer.hpp

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_ARTIFACT_LEXER_HPP
+#define MENDER_ARTIFACT_LEXER_HPP
+
+#include <cstdint>
+#include <string>
+#include <vector>
+#include <cstdlib>
+#include <cstdio>
+#include <unordered_map>
+
+#include <system_error>
+
+#include <common/io.hpp>
+#include <common/log.hpp>
+#include <common/expected.hpp>
+#include <artifact/tar/tar.hpp>
+
+#include <artifact/sha/sha.hpp>
+
+#include <artifact/token.hpp>
+
+
+namespace mender {
+namespace artifact {
+namespace lexer {
+
+using namespace std;
+
+namespace log = mender::common::log;
+
+class Lexer {
+private:
+	std::shared_ptr<mender::tar::Reader> &tar_reader;
+
+public:
+	token::Token current;
+
+	Lexer(std::shared_ptr<mender::tar::Reader> &tr) :
```

Comment:
Semantically it makes sense, but I'm not sure if it's a good idea since we often don't know where a `Reader` comes from. Providers of `Reader`s will tend to provide it as `shared_ptr` to be as flexible as possible, and we can't convert a `shared_ptr` to a `unique_ptr`. For example, this is used [in the HTTP API](https://github.com/mendersoftware/mender/blob/ea30d51e0fee413a181dbe69a21d962f5dafab01/common/http.hpp#L85).